packages/vite/src/node/server/middlewares/static.ts

```diff
@@ -106,15 +109,32 @@ export function serveRawFsMiddleware(
   }
 }
 
-export function ensureServingAccess(url: string, serveRoot: string): void {
+export function ensureServingAccess(
+  url: string,
+  { root, strict }: Required<FileSystemServeOptions>,
+  logger: Logger
+): void {
+  // TODO: early return, should remove once we polished the restriction logic
+  if (!strict) return
+
   const normalizedUrl = normalizePath(url)
-  if (!normalizedUrl.startsWith(serveRoot + path.posix.sep)) {
-    throw new AccessRestrictedError(
-      `The request url "${normalizedUrl}" is outside of vite dev server root "${serveRoot}". 
-      For security concerns, accessing files outside of workspace root is restricted since Vite v2.3.x. 
-      Refer to docs https://vitejs.dev/config/#server-fsserve-root for configurations and more details.`,
-      normalizedUrl,
-      serveRoot
-    )
+  if (!normalizedUrl.startsWith(root + path.posix.sep)) {
+    if (strict) {
+      throw new AccessRestrictedError(
+        `The request url "${normalizedUrl}" is outside of vite dev server root "${root}". 
+        For security concerns, accessing files outside of workspace root is restricted since Vite v2.3.x. 
+        Refer to docs https://vitejs.dev/config/#server-fsserve-root for configurations and more details.`,
+        normalizedUrl,
+        root
+      )
+    } else {
+      // TODO: warn for potential unrestricted access
+      logger.warnOnce(
+        `For security concerns, accessing files outside of workspace root will ` +
+          `be restricted by default in the future version of Vite. ` +
+          `Refer to [] for more`
```

Comment:
`Refer to [] for more` ❓ To what?

Reply:
It will not be triggered as this moment, left as a placeholder for the future.